### PR TITLE
Optimize `eoRanking`: Add caching and index vector

### PR DESCRIPTION
**Pull request by Alessandro624**
This PR improves the performance of the eoRanking class by introducing two key enhancements:

- Cached Coefficients:

    - Precomputes and stores intermediate constants (e.g., alpha, beta, gamma).

    - These are recomputed only when the population size changes, improving runtime efficiency for repeated evaluations.

- Index Vector:

    - Replaces the costly lookfor loop with a std::vector for O(1) lookup of individual indices.

    - This avoids potential runtime errors and simplifies the ranking logic.

Fully preserves compatibility with the EO framework (eoPerf2Worth, eoPop).

### CODE REVIEW

**Review by nojhan — CHANGES_REQUESTED** [review]
Thanks for the contribution! This is a neat feature, which we must include in Paradiseo.

However, adding a cache system rely on the idea that the instantiated operator is always used in the same context. In the case of automated design, this is not always the case. Thus, I think this class should either:
- be a new class, like `eoRankingCached`,
- have a configurable cache, with the cache disabled by default.

I would also like to have more comments, given that the system is not immediately obvious to the reader. Something like starting from the PR message and extend a bit, in the class' header, perhaps?

Finally, a simple test ensuring that both implementations produces the same result would be a convincing check that everything is perfectly working.

**Comment by Alessandro624** [comment]
Thank you for your review and suggestions. I've implemented the requested changes:

- **Created a separate `eoRankingCached` class** to maintain clear separation from the original implementation

- **Added more comments** including:

    - Detailed class header comments explaining the caching mechanism

    - Usage guidelines

    - Clear warning about constant population size requirement

- **Implemented some tests** that verifies:

    - Identical results between cached and original versions

    - Edge case handling (including minimum population size)

File: eo/src/eoRanking.h [hunk 1]
```diff
@@ -40,73 +40,175 @@ template <class EOT>
...
-    using eoPerf2Worth<EOT>::value;
+  using eoPerf2Worth<EOT>::value;
 
   /* Ctor:
    @param _p selective pressure (in (1,2]
    @param _e exponent (1 == linear)
   */
-  eoRanking(double _p=2.0, double _e=1.0):
-    pressure(_p), exponent(_e) {}
+  eoRanking(double _p = 2.0, double _e = 1.0) : pressure(_p), exponent(_e) {}
```
Comment (nojhan) [inline eo/src/eoRanking.h:49]: May I ask you to use this occasion to improve a little bit of the old code ? :-)
```suggestion
  eoRanking(double _p = 2.0, double _e = 1.0) : pressure(_p), exponent(_e) {
      assert(1 < pressure and exponent <= 2);
  }
```

File: eo/src/eoRanking.h [hunk 2]
```diff
@@ -40,73 +40,175 @@ template <class EOT>
...
+ * This class implements the same ranking algorithm as eoRanking but adds a caching layer
+ * that stores frequently used values when the population size remains constant between
+ * calls. This optimization is particularly useful in steady-state evolution where the
+ * population size typically doesn't change between selection operations.
+ *
+ * The caching mechanism stores:
+ * - Population size related values (pSize, pSizeMinusOne)
+ * - Precomputed coefficients (alpha, beta, gamma)
+ *
+ * Note: This optimization should only be used when the population size remains constant
```
Comment (nojhan) [inline eo/src/eoRanking.h:122]: Doxygen allows making such a message more visible:
```suggestion
 * @warning This optimization should only be used when the population size remains constant
```

File: eo/src/eoRanking.h [hunk 3]
```diff
@@ -40,73 +40,175 @@ template <class EOT>
...
+{
+public:
+  using eoPerf2Worth<EOT>::value;
 
+  /* Ctor:
+   @param _p selective pressure (in (1,2]
+   @param _e exponent (1 == linear)
+  */
+  eoRankingCached(double _p = 2.0, double _e = 1.0)
+      : pressure(_p), exponent(_e), cached_pSize(0) {}
```
Comment (nojhan) [inline eo/src/eoRanking.h:138]:
```suggestion
      : pressure(_p), exponent(_e), cached_pSize(0) {
          assert(1 < pressure and pressure <= 2);
      }
```

File: eo/test/t-eoRankingCached.cpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+class RankingTest
+{
+public:
+    RankingTest(eoParser &parser, eoEvalFuncCounter<eoReal<double>> &_eval, unsigned size = 100) : rng(0),
+                                                                                                   popSize(size),
+                                                                                                   seedParam(parser.createParam(uint32_t(time(0)), "seed", "Random seed", 'S')),
+                                                                                                   pressureParam(parser.createParam(1.5, "pressure", "Selective pressure", 'p')),
+                                                                                                   exponentParam(parser.createParam(1.0, "exponent", "Ranking exponent", 'e')),
+                                                                                                   eval(_eval)
```
Comment (nojhan) [inline eo/test/t-eoRankingCached.cpp:16]:
```suggestion
        popSize(size),
        seedParam(parser.createParam(uint32_t(time(0)), "seed", "Random seed", 'S')),
        pressureParam(parser.createParam(1.5, "pressure", "Selective pressure", 'p')),
        exponentParam(parser.createParam(1.0, "exponent", "Ranking exponent", 'e')),
        eval(_eval)
```

File: eo/test/t-eoRankingCached.cpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    const std::vector<double> &cachedValues = rankingCached.value();
+
+    for (unsigned i = 0; i < fixture.pop.size(); ++i)
+    {
+        if (abs(values[i] - cachedValues[i]) > 1e-9)
+        {
+            throw std::runtime_error("Inconsistent ranking values between implementations");
+        }
+    }
+    std::cout << "Test 1 passed: Both implementations produce identical results\n";
```
Comment (nojhan) [inline eo/test/t-eoRankingCached.cpp:71]: `clog` is more consistent with logging, and `endl` allows for flushing.
```suggestion
    std::clog << "Test 1 passed: Both implementations produce identical results" << std::endl;
```

File: eo/test/t-eoRankingCached.cpp [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+    ranking(smallPop);
+    rankingCached(smallPop);
+
+    if (ranking.value()[0] >= ranking.value()[1] ||
+        rankingCached.value()[0] >= rankingCached.value()[1])
+    {
+        throw std::runtime_error("Invalid ranking for population size 2");
+    }
+    std::cout << "Test 2 passed: Minimum population size handled correctly\n";
```
Comment (nojhan) [inline eo/test/t-eoRankingCached.cpp:102]:
```suggestion
    std::clog << "Test 2 passed: Minimum population size handled correctly" << std::endl;
```

File: eo/test/t-eoRankingCached.cpp [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+    eoRankingCached<eoReal<double>> rankingCached(fixture.pressure(), fixture.exponent());
+
+    // First run - should compute all values
+    rankingCached(fixture.pop);
+    const auto firstValues = rankingCached.value();
+
+    // Modify fitness values but keep same population size
+    for (auto &ind : fixture.pop)
+        ind[0] = fixture.rng.uniform();
```
Comment (nojhan) [inline eo/test/t-eoRankingCached.cpp:120]:
```suggestion
    for (auto &ind : fixture.pop) {
        ind[0] = fixture.rng.uniform();
    }
```

File: eo/test/t-eoRankingCached.cpp [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    newInd.resize(1);
+    newInd[0] = fixture.rng.uniform();
+    fixture.pop.push_back(newInd);
+
+    apply<eoReal<double>>(eval, fixture.pop);
+
+    // Third run - should recompute coefficients
+    rankingCached(fixture.pop);
+
+    std::cout << "Test 3 passed: Caching mechanism properly invalidated\n";
```
Comment (nojhan) [inline eo/test/t-eoRankingCached.cpp:138]:
```suggestion
    std::clog << "Test 3 passed: Caching mechanism properly invalidated" << std::endl;
```

File: eo/test/t-eoRankingCached.cpp [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    {
+        eoParser parser(argc, argv);
+        test_Consistency(parser);
+        test_MinPopulationSize(parser);
+        test_CachingEffectiveness(parser);
+        return 0;
+    }
+    catch (std::exception &e)
+    {
+        std::cout << "Exception: " << e.what() << std::endl;
```
Comment (nojhan) [inline eo/test/t-eoRankingCached.cpp:153]:
```suggestion
        std::clog << "Exception: " << e.what() << std::endl;
```

**Review by nojhan — CHANGES_REQUESTED** [review]
Impressive work!

Apart from the very minor comments, I think it would be better if we had those two classes in two separate files. I know it's not always the case in all the old codebase, but I think that would really help discoverability, for users who look at files directly.

Also, the indentation seems to be inconsistent, with sometime two spaces, sometime four. I think four is more or less the standard in Paradiseo (even if we have mixed styles…).

**Comment by Alessandro624** [comment]
Thank you for your review and valuable feedback. I have implemented both of your requested changes:

- Separated the classes into individual files to improve code organization and discoverability

- Standardized the indentation to use 4 spaces consistently throughout

Additionally, the newly added test cases confirm that all assertions are functioning as intended.

File: eo/src/eoRanking.h [hunk 10]
```diff
@@ -40,73 +40,85 @@ template <class EOT>
...
+     @param _e exponent (1 == linear)
+    */
+    eoRanking(double _p = 2.0, double _e = 1.0) : pressure(_p), exponent(_e)
+    {
+        if (pressure <= 1.0)
+        {
+            std::string msg = "eoRanking: pressure must be > 1.0";
+            eo::log << eo::errors << "ERROR: " << msg << std::endl;
+            throw eoException(msg);
+        }
```
Comment (nojhan) [inline eo/src/eoRanking.h:56]: I understand the idea of having a consistent errors managements. However, since we are paying attention to performances, we don't want to add too many tests for states that will never occur in experiment conditions. Thus, parameters that should be all set up should be tested with `assert`, which is declutched in `Debug` builds, saving (overall *a lot* of) tests.

Exceptions are to be used for checking states that we cannot manage easily from high-level interfaces, for instance variables that changes during computations.

File: eo/src/eoRanking.h [hunk 11]
```diff
@@ -40,73 +40,85 @@ template <class EOT>
...
 
-  /* helper function: finds index in _pop of _eo, an EOT * */
-  int lookfor(const EOT *_eo, const eoPop<EOT>& _pop)
+        if (exponent > 2.0)
+        {
+            std::string msg = "eoRanking: exponent must be <= 2.0";
+            eo::log << eo::errors << "ERROR: " << msg << std::endl;
+            throw eoException(msg);
+        }
```
Comment (nojhan) [inline eo/src/eoRanking.h:63]: idem

File: eo/src/eoRanking.h [hunk 12]
```diff
@@ -40,73 +40,85 @@ template <class EOT>
...
-      unsigned int pSizeMinusOne = pSize-1;
+        std::vector<const EOT *> rank;
+        _pop.sort(rank);
+        unsigned pSize = _pop.size();
+        unsigned int pSizeMinusOne = pSize - 1;
 
-      if (pSize <= 1)
-        throw eoPopSizeException(pSize,"cannot do ranking with population of size <= 1");
+        if (pSize <= 1)
+            throw eoPopSizeException(pSize, "cannot do ranking with population of size <= 1");
```
Comment (nojhan) [inline eo/src/eoRanking.h:90]: This one is in the grey zone, for instance, given that some algorithms have the population size changing…

File: eo/src/eoRankingCached.h [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+            eo::log << eo::errors << "ERROR: " << msg << std::endl;
+            throw eoException(msg);
+        }
+
+        if (exponent > 2.0)
+        {
+            std::string msg = "eoRankingCached: exponent must be <= 2.0";
+            eo::log << eo::errors << "ERROR: " << msg << std::endl;
+            throw eoException(msg);
+        }
```
Comment (nojhan) [inline eo/src/eoRankingCached.h:74]: idem, about asserts.

**Review by nojhan — CHANGES_REQUESTED** [review]
Almost perfect! I'm just nitpicking about asserts, but once fixed, I will merge. I tested locally, and everything merges well and passes tests.

**Comment by nojhan** [comment]
Also, don't forget to "resolve conversations" that are fixed in further commits.

File: eo/src/eoRanking.h [hunk 14]
```diff
@@ -48,19 +48,7 @@ class eoRanking : public eoPerf2Worth<EOT> // false: do not cache fitness
...
-            throw eoException(msg);
-        }
-
-        if (exponent > 2.0)
-        {
-            std::string msg = "eoRanking: exponent must be <= 2.0";
-            eo::log << eo::errors << "ERROR: " << msg << std::endl;
-            throw eoException(msg);
-        }
+        assert(1 < pressure and exponent <= 2);
```
Comment (nojhan) [inline eo/src/eoRanking.h:51]: Doc says "pressure in (1,2]", but assert says something different. Which one is right?

**Review by nojhan — CHANGES_REQUESTED** [review]
Just a last inconsistency.

**Comment by Alessandro624** [comment]
Thanks for the detailed feedback and suggestions!

**Comment by nojhan** [comment]
No problem. I'm looking forward for the next contribution ;-) And I hope you find Paradiseo useful.
